crates/circuit/src/dag_circuit.rs

```diff
@@ -4504,6 +4436,121 @@ impl DAGCircuit {
+                        match type_ {
+                            DAGStretchType::Capture => {
+                                target_dag.add_captured_stretch(stretch)?;
+                            }
+                            DAGStretchType::Declare => {
+                                target_dag.add_declared_stretch(stretch)?;
+                            }
+                        }
+                    }
+                    DAGIdentifierInfo::Var(DAGVarInfo { var, type_, .. }) => {
+                        let var = self.vars.get(*var).unwrap().clone();
+                        target_dag.add_var(var, *type_)?;
+                    }
+                }
+            }
+        } else if vars_mode == "captures" {
+            for info in self.identifier_info.values() {
+                match info {
+                    DAGIdentifierInfo::Stretch(DAGStretchInfo { stretch, .. }) => {
+                        let stretch = self.stretches.get(*stretch).unwrap().clone();
+                        target_dag.add_captured_stretch(stretch)?;
+                    }
+                    DAGIdentifierInfo::Var(DAGVarInfo { var, .. }) => {
+                        let var = self.vars.get(*var).unwrap().clone();
+                        target_dag.add_var(var, DAGVarType::Capture)?;
+                    }
+                }
+            }
+        } else if vars_mode != "drop" {
+            return Err(PyValueError::new_err(format!(
+                "unknown vars_mode: '{}'",
+                vars_mode
+            )));
+        }
+
+        Ok(target_dag)
+    }
+
+    /// Create an empty DAG with the canonical "physical" register of the correct length, with all
+    /// classical data and metadata retained.
+    ///
+    /// This is similar to [copy_empty_like_with_capacity] with `vars_mode="alike"`, and copies the
+    /// same things over (global phase, metadata, etc) it does, except for replacing the qubits.
+    ///
+    /// This method is intended for use by passes that are converting a virtual DAG to a physical
+    /// one.
+    pub fn physical_empty_like_with_capacity(
+        &self,
+        num_qubits: usize,
+        num_ops: usize,
+        num_edges: usize,
+    ) -> PyResult<Self> {
+        // This current implementation is not particularly efficient - it's not hard to speed it up
+        // if it becomes a problem.
+        let extra_qubits = num_qubits.saturating_sub(self.num_qubits());
+        let mut out = self.copy_empty_like_with_capacity(
+            // Account for the in- and out-nodes of any qubits that we'll be adding in, which
+            // `copy_empty_like` doesn't know about.
+            num_ops + 2 * extra_qubits,
+            num_edges + extra_qubits,
+            "alike",
+        )?;
+        out.remove_qubits((0..out.num_qubits()).map(Qubit::new))?;
+        out.qargs_interner = Interner::new();
+        out.add_qreg(QuantumRegister::new_owning("q", num_qubits as u32))?;
```

Comment:
:see_no_evil:

Reply:
For better or worse, this _does_ define a physical circuit.